packages/debugger/src/panels/kernelSources/body.tsx

```diff
@@ -21,19 +23,149 @@ import { UseSignal } from '@jupyterlab/ui-components';
+ * Recursive component for rendering a tree node (directory or file).
  */
-const SOURCE_CLASS = 'jp-DebuggerKernelSource-source';
+function TreeNode(props: ITreeNodeProps): JSX.Element {
+  const { name, path, modName, children, onOpen } = props;
+  const [isOpen, setIsOpen] = useState(false);
+
+  const isDirectory = !!children && children.length > 0;
+
+  // Compute collapsed chain dynamically (only when expanded)
+  let displayName = name;
+  let displayChildren = children;
+
+  if (isOpen && isDirectory) {
+    const chain: string[] = [name];
+    let currentChildren = children;
+
+    // collapse linear folder chains dynamically
+    while (
+      currentChildren &&
+      currentChildren.length === 1 &&
+      currentChildren[0].children &&
+      currentChildren[0].children!.length > 0
+    ) {
+      const next = currentChildren[0];
+      chain.push(next.name);
+      currentChildren = next.children ?? [];
+    }
+
+    if (chain.length > 1) {
+      // Infer separator from first child's modName if available
+      const childModName = currentChildren?.[0]?.modName ?? modName;
+      const sep =
+        childModName.includes('.') && !childModName.includes('/') ? '.' : '/';
+      displayName = chain.join(sep);
+      displayChildren = currentChildren;
+    }
+  }
+
+  const handleClick = (): void => {
+    if (isDirectory) {
+      setIsOpen(!isOpen);
+    } else {
+      onOpen(path);
+    }
+  };
+
+  return (
+    <div className={isDirectory ? DIR_CLASS : SOURCE_CLASS} title={path}>
+      <div className="jp-DebuggerKernelSource-item" onClick={handleClick}>
+        <LabIcon.resolveReact
+          icon={
+            isDirectory
+              ? isOpen
+                ? caretDownIcon
+                : caretRightIcon
+              : openKernelSourceIcon
+          }
+          className={classes('jp-Icon')}
+          tag={null}
+        />
+        <span className="jp-DebuggerKernelSource-label">{displayName}</span>
+      </div>
+      {isDirectory && isOpen && displayChildren && (
+        <div className="jp-DebuggerKernelSource-children">
+          {displayChildren.map((child, i) => (
+            <TreeNode key={i} {...child} onOpen={onOpen} />
+          ))}
+        </div>
+      )}
+    </div>
+  );
+}
+
+/**
+ * Convert a flat list of kernel sources into a hierarchical directory tree.
+ */
+function buildTree(modules: IDebugger.KernelSource[]): ITreeNodeProps[] {
+  const root: Record<string, any> = {};
+
+  for (const mod of modules) {
+    const sep = mod.name.includes('.') && !mod.name.includes('/') ? '.' : '/';
+    const parts = mod.name.split(sep);
+    let current = root;
+    for (let i = 0; i < parts.length; i++) {
+      const part = parts[i];
+      if (!part) continue;
+      if (!current[part]) {
+        current[part] = {
+          __children__: {},
+          __namePath__: parts.slice(0, i + 1).join(sep),
+          __filePath__: mod.path,
+          __modName__: mod.name
```

Comment:
Is there a reason to use `__` prefix and suffix?

Reply:
The `__` prefix and suffix is intentional to avoid conflict with filenames like children, path, filePath, or moduleName. Since any module's file name can legally be anything - so this is just to be on the safer side.